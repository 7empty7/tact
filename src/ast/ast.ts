import { Interval as RawInterval, Node as RawNode } from 'ohm-js';

export class ASTRef {
    readonly #interval: RawInterval;

    constructor(interval: RawInterval) {
        this.#interval = interval;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }
}

export type ASTPrimitive = {
    kind: 'primitive',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Values
//

export type ASTNumber = {
    kind: 'number',
    id: number,
    value: bigint,
    ref: ASTRef
}

export type ASTID = {
    kind: 'id',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTBoolean = {
    kind: 'boolean',
    id: number,
    value: boolean,
    ref: ASTRef
}

export type ASTNull = {
    kind: 'null',
    id: number,
    ref: ASTRef
}

//
// Types
//

export type ASTTypeRef = {
    kind: 'type_ref',
    id: number,
    name: string,
    optional: boolean,
    ref: ASTRef
}

//
// Expressions
//

export type ASTOpBinary = {
    kind: 'op_binary',
    id: number,
    op: '+' | '-' | '*' | '/' | '!=' | '>' | '<' | '>=' | '<=' | '==' | '&&' | '||',
    left: ASTExpression,
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpUnary = {
    kind: 'op_unary',
    id: number,
    op: '+' | '-' | '!' | '!!',
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpField = {
    kind: 'op_field'
    id: number,
    src: ASTExpression,
    name: string,
    ref: ASTRef
}

export type ASTOpCall = {
    kind: 'op_call'
    id: number,
    src: ASTExpression,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpCallStatic = {
    kind: 'op_static_call'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpNew = {
    kind: 'op_new'
    id: number,
    type: string,
    args: ASTNewParameter[],
    ref: ASTRef
}

export type ASTNewParameter = {
    kind: 'new_parameter'
    id: number,
    name: string,
    exp: ASTExpression,
    ref: ASTRef
}

//
// Program
//

export type ASTProgram = {
    kind: 'program',
    id: number,
    entries: (ASTStruct | ASTContract | ASTPrimitive | ASTFunction | ASTNativeFunction)[]
}

export type ASTStruct = {
    kind: 'def_struct',
    id: number,
    name: string
    fields: ASTField[],
    ref: ASTRef
}

export type ASTField = {
    kind: 'def_field',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTContract = {
    kind: 'def_contract',
    id: number,
    name: string,
    declarations: (ASTField | ASTFunction)[],
    ref: ASTRef
}

export type ASTArgument = {
    kind: 'def_argument',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTFunctionAttribute = { type: 'public' } | { type: 'get' };

export type ASTFunction = {
    kind: 'def_function',
    id: number,
    attribute: ASTFunctionAttribute[],
    name: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTNativeFunction = {
    kind: 'def_native_function',
    id: number,
    name: string,
    nativeName: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    ref: ASTRef
}

//
// Statements
//

export type ASTStatementLet = {
    kind: 'statement_let',
    id: number,
    name: string,
    type: ASTTypeRef,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementReturn = {
    kind: 'statement_return',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementCall = {
    kind: 'statement_call',
    id: number,
    expression: ASTOpCall | ASTOpCallStatic,
    ref: ASTRef
}

export type ASTSTatementAssign = {
    kind: 'statement_assign',
    id: number,
    path: string[],
    expression: ASTExpression,
    ref: ASTRef
}

//
// Unions
//

export type ASTStatement = ASTStatementLet | ASTStatementReturn | ASTStatementCall | ASTSTatementAssign;
export type ASTExpression = ASTOpBinary | ASTOpUnary | ASTOpField | ASTNumber | ASTID | ASTBoolean | ASTOpCall | ASTOpCallStatic | ASTOpNew;
export type ASTNode = ASTExpression | ASTProgram | ASTStruct | ASTField | ASTContract | ASTArgument | ASTFunction | ASTOpCall | ASTStatementLet | ASTStatementReturn | ASTProgram | ASTPrimitive | ASTOpCallStatic | ASTStatementCall | ASTNativeFunction | ASTSTatementAssign | ASTOpNew | ASTNewParameter | ASTTypeRef;
export type ASTType = ASTPrimitive | ASTStruct | ASTContract;

export function isStatement(src: ASTNode): src is ASTStatement {
    return src.kind === 'statement_let' || src.kind === 'statement_return' || src.kind === 'statement_call' || src.kind === 'statement_assign';
}

export function isExpression(src: ASTNode): src is ASTExpression {
    return src.kind === 'op_binary' || src.kind === 'op_unary' || src.kind === 'op_field' || src.kind === 'number' || src.kind === 'id' || src.kind === 'boolean' || src.kind === 'op_new' || src.kind === 'op_call' || src.kind === 'op_static_call';
}


type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;
let nextId = 1;
export function createNode(src: DistributiveOmit<ASTNode, 'id'>): ASTNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function createRef(s: RawNode, ...extra: RawNode[]): ASTRef {
    let i = s.source;
    if (extra.length > 0) {
        i = i.coverageWith(...extra.map((e) => e.source));
    }
    return new ASTRef(i);
}

export function throwError(message: string, ref: ASTRef): never {
    throw new Error(ref.interval.getLineAndColumnMessage() + message);
}