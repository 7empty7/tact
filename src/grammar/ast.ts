import { Interval as RawInterval, Node as RawNode } from 'ohm-js';

export class ASTRef {

    static merge(...refs: ASTRef[]) {
        if (refs.length === 0) {
            throw Error('Cannot merge 0 refs');
        }
        let r = refs[0].#interval;
        for (let i = 1; i < refs.length; i++) {
            r = r.coverageWith(r, refs[i].#interval);
        }
        return new ASTRef(r);
    }

    readonly #interval: RawInterval;

    constructor(interval: RawInterval) {
        this.#interval = interval;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }
}

export type ASTPrimitive = {
    kind: 'primitive',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Values
//

export type ASTNumber = {
    kind: 'number',
    id: number,
    value: bigint,
    ref: ASTRef
}

export type ASTID = {
    kind: 'id',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTBoolean = {
    kind: 'boolean',
    id: number,
    value: boolean,
    ref: ASTRef
}

export type ASTString = {
    kind: 'string',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTNull = {
    kind: 'null',
    id: number,
    ref: ASTRef
}

export type ASTLvalueRef = {
    kind: 'lvalue_ref',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Types
//

export type ASTTypeRefSimple = {
    kind: 'type_ref_simple',
    id: number,
    name: string,
    optional: boolean,
    ref: ASTRef
}

export type ASTTypeRefMap = {
    kind: 'type_ref_map',
    id: number,
    key: string,
    value: string,
    ref: ASTRef
}

export type ASTTypeRef = ASTTypeRefSimple | ASTTypeRefMap;

//
// Expressions
//

export type ASTOpBinary = {
    kind: 'op_binary',
    id: number,
    op: '+' | '-' | '*' | '/' | '!=' | '>' | '<' | '>=' | '<=' | '==' | '&&' | '||',
    left: ASTExpression,
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpUnary = {
    kind: 'op_unary',
    id: number,
    op: '+' | '-' | '!' | '!!',
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpField = {
    kind: 'op_field'
    id: number,
    src: ASTExpression,
    name: string,
    ref: ASTRef
}

export type ASTOpCall = {
    kind: 'op_call'
    id: number,
    src: ASTExpression,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpCallStatic = {
    kind: 'op_static_call'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpNew = {
    kind: 'op_new'
    id: number,
    type: string,
    args: ASTNewParameter[],
    ref: ASTRef
}

export type ASTNewParameter = {
    kind: 'new_parameter'
    id: number,
    name: string,
    exp: ASTExpression,
    ref: ASTRef
}

//
// Program
//

export type ASTProgram = {
    kind: 'program',
    id: number,
    entries: (ASTStruct | ASTContract | ASTPrimitive | ASTFunction | ASTNativeFunction | ASTProgramImport)[]
}

export type ASTProgramImport = {
    kind: 'program_import',
    id: number,
    path: ASTString,
}

export type ASTStruct = {
    kind: 'def_struct',
    id: number,
    name: string,
    message: boolean,
    prefix: number | null,
    fields: ASTField[],
    ref: ASTRef
}

export type ASTField = {
    kind: 'def_field',
    id: number,
    name: string,
    type: ASTTypeRef,
    init: bigint | boolean | null | undefined,
    as: string | null,
    ref: ASTRef
}

export type ASTContract = {
    kind: 'def_contract',
    id: number,
    name: string,
    declarations: (ASTField | ASTFunction | ASTInitFunction | ASTReceive)[],
    ref: ASTRef
}

export type ASTArgument = {
    kind: 'def_argument',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTFunctionAttribute = { type: 'public', ref: ASTRef } | { type: 'get', ref: ASTRef } | { type: 'mutates', ref: ASTRef } | { type: 'extends', ref: ASTRef };

export type ASTFunction = {
    kind: 'def_function',
    id: number,
    attributes: ASTFunctionAttribute[],
    name: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTReceive = {
    kind: 'def_receive',
    id: number,
    selector: { kind: 'simple', arg: ASTArgument } | { kind: 'fallback' } | { kind: 'comment', comment: ASTString } | { kind: 'bounce', arg: ASTArgument },
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTNativeFunction = {
    kind: 'def_native_function',
    id: number,
    attributes: ASTFunctionAttribute[],
    name: string,
    nativeName: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    ref: ASTRef
}

export type ASTInitFunction = {
    kind: 'def_init_function',
    id: number,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

//
// Statements
//

export type ASTStatementLet = {
    kind: 'statement_let',
    id: number,
    name: string,
    type: ASTTypeRef,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementReturn = {
    kind: 'statement_return',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementExpression = {
    kind: 'statement_expression',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTSTatementAssign = {
    kind: 'statement_assign',
    id: number,
    path: ASTLvalueRef[],
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTCondition = {
    kind: 'statement_condition',
    id: number,
    expression: ASTExpression,
    trueStatements: ASTStatement[],
    falseStatements: ASTStatement[];
    elseif: ASTCondition | null,
    ref: ASTRef,
}

export type ASTStatementWhile = {
    kind: 'statement_while'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTStatementUntil = {
    kind: 'statement_until'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTStatementRepeat = {
    kind: 'statement_repeat'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}


//
// Unions
//

export type ASTStatement = ASTStatementLet | ASTStatementReturn | ASTStatementExpression | ASTSTatementAssign | ASTCondition | ASTStatementWhile | ASTStatementUntil | ASTStatementRepeat;
export type ASTExpression = ASTOpBinary | ASTOpUnary | ASTOpField | ASTNumber | ASTID | ASTBoolean | ASTOpCall | ASTOpCallStatic | ASTOpNew | ASTNull | ASTLvalueRef;
export type ASTNode = ASTExpression | ASTProgram | ASTStruct | ASTField | ASTContract | ASTArgument | ASTFunction | ASTOpCall | ASTStatementLet | ASTStatementReturn | ASTProgram | ASTPrimitive | ASTOpCallStatic | ASTStatementExpression | ASTNativeFunction | ASTSTatementAssign | ASTOpNew | ASTNewParameter | ASTTypeRef | ASTNull | ASTCondition | ASTInitFunction | ASTStatementWhile | ASTStatementUntil | ASTStatementRepeat | ASTReceive | ASTLvalueRef | ASTString | ASTProgramImport;
export type ASTType = ASTPrimitive | ASTStruct | ASTContract;

export function isStatement(src: ASTNode): src is ASTStatement {
    return src.kind === 'statement_let' || src.kind === 'statement_return' || src.kind === 'statement_expression' || src.kind === 'statement_assign' || src.kind === 'statement_condition' || src.kind === 'statement_while' || src.kind === 'statement_until' || src.kind === 'statement_repeat';
}

export function isExpression(src: ASTNode): src is ASTExpression {
    return src.kind === 'op_binary' || src.kind === 'op_unary' || src.kind === 'op_field' || src.kind === 'number' || src.kind === 'id' || src.kind === 'boolean' || src.kind === 'op_new' || src.kind === 'op_call' || src.kind === 'op_static_call';
}


type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;
let nextId = 1;
export function createNode(src: DistributiveOmit<ASTNode, 'id'>): ASTNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function __DANGER_resetNodeId() {
    nextId = 1;
}

export function createRef(s: RawNode, ...extra: RawNode[]): ASTRef {
    let i = s.source;
    if (extra.length > 0) {
        i = i.coverageWith(...extra.map((e) => e.source));
    }
    return new ASTRef(i);
}

export function throwError(message: string, ref: ASTRef): never {
    throw new Error(ref.interval.getLineAndColumnMessage() + message);
}