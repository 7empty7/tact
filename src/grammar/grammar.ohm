Tact {

    // Starting point of the program
    Program = ProgramItem*
    ProgramItem = Struct
                | Contract
                | Primitive
                | StaticFunction
                | NativeFunction

    // Built-in declarations
    Primitive = primitive Type ";"

    // Static function
    StaticFunction = Function
    NativeFunction = nameAttribute "(" id ")" native id "(" ListOf<FunctionArg,","> ")" ";" --withVoid
                   | nameAttribute "(" id ")" native id "(" ListOf<FunctionArg,","> ")" ":" Type ";" --withType
    
    // Field declarations
    Type = typeLiteral "?" --optional
         | typeLiteral --required
    Field = var id ":" Type ";"

    // Struct
    Struct = struct id "{" StructBody* "}"
    StructBody = Field

    // Contract
    Contract = contract id "{" ContractBody* "}"
    ContractBody = Field
                 | Function 

    // Function
    FunctionAttribute = public --public
                      | get    --getter
    Function = ListOf<FunctionAttribute," "> fun id "(" ListOf<FunctionArg,","> ")" "{" Statement* "}" --withVoid
             | ListOf<FunctionAttribute," "> fun id "(" ListOf<FunctionArg,","> ")" ":" Type "{" Statement* "}" --withType
             
    FunctionArg = id ":" Type

    // Statements
    Statement = StatementLet
              | StatementBlock
              | StatementReturn
              | StatementCall
              | StatementStaticCall
              | StatementAssign
    StatementBlock = "{" Statement* "}"
    StatementLet = let id ":" Type "=" Expression ";"
    StatementReturn = return Expression ";"
    StatementCall = ExpressionCall ";"
    StatementStaticCall = ExpressionStaticCall ";"
    StatementAssign = LValue Expression ";"

    // L-value
    LValue = id "=" --id
           | id "." LValue --subId

    // Expressions
    Expression = ExpressionOr
    ExpressionOr = ExpressionOr "||" ExpressionAnd --or
                 | ExpressionAnd
    ExpressionAnd = ExpressionAnd "&&" ExpressionCompare --and
                  | ExpressionCompare
    ExpressionCompare = ExpressionCompare "!=" ExpressionAdd --not
                      | ExpressionCompare "==" ExpressionAdd --eq
                      | ExpressionCompare ">" ExpressionAdd --gt
                      | ExpressionCompare ">=" ExpressionAdd --gte
                      | ExpressionCompare "<" ExpressionAdd --lt
                      | ExpressionCompare "<=" ExpressionAdd --lte
                      | ExpressionAdd
    ExpressionAdd = ExpressionAdd "+" ~"+" ExpressionMul --add
                  | ExpressionAdd "-" ~"-" ExpressionMul --sub
                  | ExpressionMul
    ExpressionMul = ExpressionMul "*" ExpressionUnary --mul
                  | ExpressionMul "/" ExpressionUnary --div
                  | ExpressionUnary
    ExpressionUnary = "-" ExpressionUnarySuffix --neg
                    | "+" ExpressionUnarySuffix --add
                    | "!" ExpressionUnarySuffix --not
                    | ExpressionUnarySuffix
    ExpressionUnarySuffix = ExpressionValue "!!" --notNull
                          | ExpressionValue
    ExpressionBracket = "(" Expression ")"

    // Order is important
    ExpressionValue = ExpressionCall
                    | ExpressionField
                    | ExpressionStaticCall
                    | ExpressionBracket
                    | ExpressionNew
                    | integerLiteral
                    | boolLiteral
                    | id
                    
    ExpressionField = ExpressionValue "." id
    ExpressionCall = ExpressionValue "." id "(" ListOf<Expression, ","> ")"
    ExpressionNew = id "{" ListOf<NewParameter, ","> "}"
    NewParameter = id ":" Expression
    ExpressionStaticCall = ExpressionValue "(" ListOf<Expression, ","> ")"

    // Type Literal
    typeLiteral = letterAsciiUC typeLiteralPart*
    typeLiteralPart = letterAscii | digit | "_"

    // Integer Literal
    // hexDigit defined in Ohm's built-in rules (otherwise: hexDigit = "0".."9" | "a".."f" | "A".."F")
    // digit defined in Ohm's built-in rules (otherwise: digit = "0".."9")
    integerLiteral = integerLiteralHex | integerLiteralDec // Order is important
    integerLiteralDec = digit+
    integerLiteralHex = "0x" hexDigit+
                      | "0X" hexDigit+

    // Letters
    letterAsciiLC = "a".."z"
    letterAsciiUC = "A".."Z"
    letterAscii = letterAsciiLC | letterAsciiUC

    // ID Literal
    idStart = letterAscii | "_"
    idPart = letterAscii | digit | "_"
    id = ~reservedWord idStart idPart*

    // Bool Literal
    boolLiteral = ("true" | "false") ~idPart

    // Keywords
    keyword = struct | contract | var | fun | let | return | primitive | extend | native | public | get | null // Order is important
    struct = "struct" ~idPart
    contract = "contract" ~idPart
    var = "var" ~idPart
    let = "let" ~idPart
    fun = "fun" ~idPart
    return = "return" ~idPart
    primitive = "primitive" ~idPart
    extend = "extend" ~idPart
    native = "native" ~idPart
    public = "public" ~idPart
    null = "null" ~idPart
    get = "get" ~idPart

    // Attributes
    nameAttribute = "@name"

    // Reserved
    reservedWord = keyword
}
