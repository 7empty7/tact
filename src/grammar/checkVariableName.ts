import { ASTRef, throwError } from "../ast/ast";

let blacklisted = [
    "for",
    "if",
    "while",
    "else",
    "return",
    "break",
    "continue",
    "switch",
    "case",
    "default",
    "function",
    "let",
    "const",
    "new",
    "self",
    "this",
    "true",
    "false",
    "null",
    "undefined",
    "class",
    "extends",
    "super",
    "import",
    "export",
    "from",
    "as",
    "try",
    "catch",
    "finally",
    "throw",
    "in",
    "of",
    "typeof",
    "instanceof",
    "void",
    "delete",
    "with",
    "debugger",
    "yield",
    "await",
    "async",
    "enum",
    "static",
    "get",
    "set",
    "cons"
];

export function checkVariableName(name: string, ref: ASTRef) {
    if (name.startsWith('__gen')) {
        throwError(`Variable name cannot start with "__gen"`, ref);
    }
    if (name.startsWith('__tact')) {
        throwError(`Variable name cannot start with "__tact"`, ref);
    }
    if (blacklisted.includes(name)) {
        throwError(`Variable name cannot be a reserved word`, ref);
    }
}