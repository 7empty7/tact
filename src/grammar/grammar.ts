import rawGrammar from './grammar.ohm-bundle';
import { ASTNode, ASTProgram, createNode, createRef } from '../ast/ast';


// Semantics
const semantics = rawGrammar.createSemantics();

// Resolve program
semantics.addOperation<ASTNode>('resolve_program', {
    Program(arg0) {
        return createNode({
            kind: 'program',
            entries: arg0.children.map((v) => v.resolve_program_item())
        });
    },
});

// Resolve program items
semantics.addOperation<ASTNode>('resolve_program_item', {
    Primitive(arg0, arg1, arg2) {
        return createNode({
            kind: 'primitive',
            name: arg1.sourceString,
            ref: createRef(this)
        });
    },
    Struct(arg0, arg1, arg2, arg3, arg4) {
        return createNode({
            kind: 'def_struct',
            name: arg1.sourceString,
            fields: arg3.children.map((v) => v.resolve_declaration()),
            ref: createRef(this)
        })
    },
    Contract(arg0, arg1, arg2, arg3, arg4) {
        return createNode({
            kind: 'def_contract',
            name: arg1.sourceString,
            declarations: arg3.children.map((v) => v.resolve_declaration()),
            ref: createRef(this)
        })
    },
    StaticFunction(arg0) {
        return arg0.resolve_declaration();
    },
    NativeFunction(arg0) {
        return arg0.resolve_declaration();
    },
});

// Struct and class declarations
semantics.addOperation<ASTNode>('resolve_declaration', {
    Field(arg0, arg1, arg2, arg3, arg4) {
        return createNode({
            kind: 'def_field',
            name: arg1.sourceString,
            type: arg3.sourceString,
            ref: createRef(this)
        })
    },
    FunctionArg(arg0, arg1, arg2) {
        return createNode({
            kind: 'def_argument',
            name: arg0.sourceString,
            type: arg2.sourceString,
            ref: createRef(this)
        })
    },
    Function_withType(arg0, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9) {
        return createNode({
            kind: 'def_function',
            name: arg1.sourceString,
            return: arg6.sourceString,
            args: arg3.asIteration().children.map((v: any) => v.resolve_declaration()),
            statements: arg8.children.map((v: any) => v.resolve_statement()),
            ref: createRef(this)
        })
    },
    Function_withVoid(arg0, arg1, arg2, arg3, arg4, arg5, arg6, arg7) {
        return createNode({
            kind: 'def_function',
            name: arg1.sourceString,
            return: null,
            args: arg3.asIteration().children.map((v: any) => v.resolve_declaration()),
            statements: arg6.children.map((v: any) => v.resolve_statement()),
            ref: createRef(this)
        })
    },
    NativeFunction_withType(arg0, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9, arg10, arg11) {
        return createNode({
            kind: 'def_native_function',
            name: arg5.sourceString,
            nativeName: arg2.sourceString,
            return: arg10.sourceString,
            args: arg7.asIteration().children.map((v: any) => v.resolve_declaration()),
            ref: createRef(this)
        })
    },
    NativeFunction_withVoid(arg0, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9) {
        return createNode({
            kind: 'def_native_function',
            name: arg5.sourceString,
            nativeName: arg2.sourceString,
            return: null,
            args: arg7.asIteration().children.map((v: any) => v.resolve_declaration()),
            ref: createRef(this)
        })
    },
});

// Statements
semantics.addOperation<ASTNode>('resolve_statement', {
    StatementLet(arg0, arg1, arg2, arg3, arg4, arg5, arg6) {
        return createNode({
            kind: 'statement_let',
            name: arg1.sourceString,
            type: arg3.sourceString,
            expression: arg5.resolve_expression(),
            ref: createRef(this)
        })
    },
    StatementReturn(arg0, arg1, arg2) {
        return createNode({
            kind: 'statement_return',
            expression: arg1.resolve_expression(),
            ref: createRef(this)
        })
    },
    StatementCall(arg0, arg1) {
        return createNode({
            kind: 'statement_call',
            expression: arg0.resolve_expression(),
            ref: createRef(this)
        })
    },
    StatementStaticCall(arg0, arg1) {
        return createNode({
            kind: 'statement_call',
            expression: arg0.resolve_expression(),
            ref: createRef(this)
        })
    },
    StatementAssign(arg0, arg1, arg2) {
        return createNode({
            kind: 'statement_assign',
            path: arg0.resolve_lvalue(),
            expression: arg1.resolve_expression(),
            ref: createRef(this)
        })
    }
});

// LValue
semantics.addOperation<string[]>('resolve_lvalue', {
    LValue_id(arg0, arg1) {
        return [arg0.sourceString];
    },
    LValue_subId(arg0, arg1, arg2) {
        return [arg0.sourceString, ...arg2.resolve_lvalue()];
    }
})

// Expressions
semantics.addOperation<ASTNode>('resolve_expression', {

    // Literals
    integerLiteral(n) {
        return createNode({ kind: 'number', value: BigInt(n.sourceString), ref: createRef(this) }); // Parses dec-based integer and hex-based integers
    },
    boolLiteral(arg0) {
        return createNode({ kind: 'boolean', value: arg0.sourceString === 'true', ref: createRef(this) });
    },
    id(arg0, arg1) {
        return createNode({ kind: 'id', value: arg0.sourceString + arg1.sourceString, ref: createRef(this) });
    },

    // Binary
    ExpressionAdd_add(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '+', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionAdd_sub(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '-', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionMul_div(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '/', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionMul_mul(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '*', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionCompare_eq(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '==', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionCompare_not(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '!=', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionCompare_gt(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '>', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionCompare_gte(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '>=', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionCompare_lt(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '>', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionCompare_lte(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '>=', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionOr_or(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '||', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },
    ExpressionAnd_and(arg0, arg1, arg2) {
        return createNode({ kind: 'op_binary', op: '&&', left: arg0.resolve_expression(), right: arg2.resolve_expression(), ref: createRef(this) });
    },

    // Unary
    ExpressionUnary_add(arg0, arg1) {
        return createNode({ kind: 'op_unary', op: '+', right: arg1.resolve_expression(), ref: createRef(this) });
    },
    ExpressionUnary_neg(arg0, arg1) {
        return createNode({ kind: 'op_unary', op: '-', right: arg1.resolve_expression(), ref: createRef(this) });
    },
    ExpressionUnary_not(arg0, arg1) {
        return createNode({ kind: 'op_unary', op: '!', right: arg1.resolve_expression(), ref: createRef(this) });
    },
    ExpressionBracket(arg0, arg1, arg2) {
        return arg1.resolve_expression();
    },

    // Access
    ExpressionField(arg0, arg1, arg2) {
        return createNode({ kind: 'op_field', src: arg0.resolve_expression(), name: arg2.sourceString, ref: createRef(this) });
    },
    ExpressionCall(arg0, arg1, arg2, arg3, arg4, arg5) {
        return createNode({ kind: 'op_call', src: arg0.resolve_expression(), name: arg2.sourceString, args: arg4.asIteration().children.map((v: any) => v.resolve_expression()), ref: createRef(this) });
    },
    ExpressionStaticCall(arg0, arg1, arg2, arg3) {
        return createNode({ kind: 'op_static_call', name: arg0.sourceString, args: arg2.asIteration().children.map((v: any) => v.resolve_expression()), ref: createRef(this) });
    },
});

export function parse(src: string): ASTProgram {
    let matchResult = rawGrammar.match(src);
    if (matchResult.failed()) {
        throw new Error(matchResult.message);
    }
    let res = semantics(matchResult).resolve_program();
    return res;
}