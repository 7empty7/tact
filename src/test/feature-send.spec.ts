import { toNano } from 'ton-core';
import { ContractSystem } from '@tact-lang/emulator';
import { __DANGER_resetNodeId } from '../grammar/ast';
import { SendTester } from './features/output/send_SendTester';

describe('feature-send', () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    it('should send reply correctly', async () => {

        // Init
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure');
        let contract = system.open(await SendTester.fromInit());
        let tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano('10') }, { $$type: 'Deploy', queryId: 0n });
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();

        await contract.send(treasure, { value: toNano('10') }, 'Hello');
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
    });
});