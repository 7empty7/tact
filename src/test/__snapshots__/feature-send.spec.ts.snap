// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feature-send should send reply correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{946A98B60000000000000000}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQDrQ0t-d2wKLVVkO8qNf1sP9nYix-YDncW5E6YuH_nm2Umm",
          "type": "internal",
          "value": 10000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8598n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Deploy success",
              "type": "text",
            },
            "to": null,
            "type": "external-out",
          },
        ],
      },
    ],
  },
]
`;

exports[`feature-send should send reply correctly 2`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Hello",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQDrQ0t-d2wKLVVkO8qNf1sP9nYix-YDncW5E6YuH_nm2Umm",
          "type": "internal",
          "value": 10000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8480n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "World",
              "type": "text",
            },
            "bounce": true,
            "from": "kQDrQ0t-d2wKLVVkO8qNf1sP9nYix-YDncW5E6YuH_nm2Umm",
            "to": "@treasure(treasure)",
            "type": "internal",
            "value": 9990348000n,
          },
        ],
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Something",
              "type": "text",
            },
            "to": null,
            "type": "external-out",
          },
        ],
      },
    ],
  },
]
`;
